flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/strategy/AdaptiveSkewedJoinOptimizationStrategy.java

```diff
@@ -301,22 +301,23 @@ private static boolean existBytesLargerThanThreshold(long[] inputBytes, long thr
     }
 
     private static boolean canPerformOptimizationAutomatic(ImmutableStreamNode adaptiveJoinNode) {
-        // In AUTO mode, we need to ensure that there are no intra-correlated out edge to ensure the
-        // application of this optimization wouldn't break data correctness or introduce additional
-        // performance overhead.
+        // In AUTO mode, we need to ensure that all pointwise out edges do not have
+        // intra-correlation to ensure the application of this optimization wouldn't break data
+        // correctness or introduce additional performance overhead.
         return adaptiveJoinNode.getOutEdges().stream()
-                .noneMatch(ImmutableStreamEdge::isIntraInputKeyCorrelated);
+                .noneMatch(edge -> edge.isPointwiseEdge() && edge.isIntraInputKeyCorrelated());
```

Comment:
Could we introduce a whitelist to check whether the output partitioner can maintain key group correlation by itself, such as with Hash, Custom, etc.?